rust/lance-table/src/rowids.rs

```diff
@@ -295,6 +298,142 @@ impl RowIdSequence {
         }
         None
     }
+
+    /// Given a mask of row ids, calculate the offset ranges of the row ids that are present
+    /// in the sequence.
+    ///
+    /// For example, given a mask that selects all even ids and a sequence that is
+    /// [80..85, 86..90, 14]
+    ///
+    /// this will return [0, 2, 4, 5, 7, 9]
+    /// because the range expands to
+    ///
+    /// [80, 81, 82, 83, 84, 86, 87, 88, 89, 14] with offsets
+    /// [ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9]
+    ///
+    /// This function is useful when determining which row offsets to read from a fragment given
+    /// a mask.
+    pub fn mask_to_offset_ranges(&self, mask: &RowIdMask) -> Vec<Range<u64>> {
```

Comment:
This is a rather bespoke function, so I wasn't sure if it belonged here or as a helper method in the exec node itself.  However, it can stand alone here so I left it here.

It could also possibly be implemented by first applying a mask to the `RowIdSequence` to get a "selection vector" and then materializing the selection vector into offsets.  However, that seemed like more work and I don't know if it would be any easier.

In fact, `Vec<Range<u64>>` _is_ a sort of "compressed selection vector".